library/core/tests/fmt/num.rs

```diff
@@ -128,28 +128,370 @@ fn test_format_int_exp_precision() {
+    assert_eq!(format!("{:.2e}", -1085), "-1.08e3");
+    assert_eq!(format!("{:.2e}", -1065), "-1.06e3");
+    assert_eq!(format!("{:.2e}", -1045), "-1.04e3");
+    assert_eq!(format!("{:.2e}", -1025), "-1.02e3");
+    assert_eq!(format!("{:.2e}", -1005), "-1.00e3");
+    assert_eq!(format!("{:.2e}",  1005), "1.00e3" );
+    assert_eq!(format!("{:.2e}",  1025), "1.02e3" );
+    assert_eq!(format!("{:.2e}",  1045), "1.04e3" );
+    assert_eq!(format!("{:.2e}",  1065), "1.06e3" );
+    assert_eq!(format!("{:.2e}",  1085), "1.08e3" );
+    assert_eq!(format!("{:.2e}",  1105), "1.10e3" );
+    assert_eq!(format!("{:.2e}",  1125), "1.12e3" );
+    assert_eq!(format!("{:.2e}",  1145), "1.14e3" );
+    assert_eq!(format!("{:.2e}",  1165), "1.16e3" );
+    assert_eq!(format!("{:.2e}",  1185), "1.18e3" );
+    assert_eq!(format!("{:.2e}",  1205), "1.20e3" );
+    assert_eq!(format!("{:.2e}",  1225), "1.22e3" );
+    assert_eq!(format!("{:.2e}",  1245), "1.24e3" );
+    assert_eq!(format!("{:.2e}",  1265), "1.26e3" );
+    assert_eq!(format!("{:.2e}",  1285), "1.28e3" );
+    assert_eq!(format!("{:.2e}",  1305), "1.30e3" );
+    assert_eq!(format!("{:.2e}",  1325), "1.32e3" );
+    assert_eq!(format!("{:.2e}",  1345), "1.34e3" );
+    assert_eq!(format!("{:.2e}",  1365), "1.36e3" );
+    assert_eq!(format!("{:.2e}",  1385), "1.38e3" );
+    assert_eq!(format!("{:.2e}",  1405), "1.40e3" );
+    assert_eq!(format!("{:.2e}",  1425), "1.42e3" );
+    assert_eq!(format!("{:.2e}",  1445), "1.44e3" );
+    assert_eq!(format!("{:.2e}",  1465), "1.46e3" );
+    assert_eq!(format!("{:.2e}",  1485), "1.48e3" );
+    assert_eq!(format!("{:.2e}",  1505), "1.50e3" );
+    assert_eq!(format!("{:.2e}",  1525), "1.52e3" );
+    assert_eq!(format!("{:.2e}",  1545), "1.54e3" );
+    assert_eq!(format!("{:.2e}",  1565), "1.56e3" );
+    assert_eq!(format!("{:.2e}",  1585), "1.58e3" );
+    assert_eq!(format!("{:.2e}",  1605), "1.60e3" );
+    assert_eq!(format!("{:.2e}",  1625), "1.62e3" );
+    assert_eq!(format!("{:.2e}",  1645), "1.64e3" );
+    assert_eq!(format!("{:.2e}",  1665), "1.66e3" );
+    assert_eq!(format!("{:.2e}",  1685), "1.68e3" );
+    assert_eq!(format!("{:.2e}",  1705), "1.70e3" );
+    assert_eq!(format!("{:.2e}",  1725), "1.72e3" );
+    assert_eq!(format!("{:.2e}",  1745), "1.74e3" );
+    assert_eq!(format!("{:.2e}",  1765), "1.76e3" );
+    assert_eq!(format!("{:.2e}",  1785), "1.78e3" );
+    assert_eq!(format!("{:.2e}",  1805), "1.80e3" );
+    assert_eq!(format!("{:.2e}",  1825), "1.82e3" );
+    assert_eq!(format!("{:.2e}",  1845), "1.84e3" );
+    assert_eq!(format!("{:.2e}",  1865), "1.86e3" );
+    assert_eq!(format!("{:.2e}",  1885), "1.88e3" );
+    assert_eq!(format!("{:.2e}",  1905), "1.90e3" );
+    assert_eq!(format!("{:.2e}",  1925), "1.92e3" );
+    assert_eq!(format!("{:.2e}",  1945), "1.94e3" );
+    assert_eq!(format!("{:.2e}",  1965), "1.96e3" );
+    assert_eq!(format!("{:.2e}",  1985), "1.98e3" );
```

Comment:
The more I look at this giant list of tests, the less I'm sure of its value. There are certainly some edge cases in here that we want to keep, but many are duplicating the exact same conditions. This makes it hard to see whether we actually have a test corresponding to the different pieces of the implementation.

For example, is there a test that exercises the power-of-10 overflow that we checked last time? I'm not sure.

Can we narrow it down to fewer targeted tests around different boundary conditions?

Reply:
Another possibility could be an exhaustive test comparing to the independent floating point implementation. Like, check *all* `i16` values compared to the same `f32` value, formatted to a few different levels of precision.

Reply:
```rust
    for i in i16::MIN..=i16::MAX {
        for p in 0..=5 {
            assert_eq!(
                format!("{i:.p$e}"),
                format!("{:.p$e}", f32::from(i)),
                "integer {i} at precision {p}"
            );
        }
    }
```